src/main/java/org/ibp/api/java/impl/middleware/study/StudyInstanceServiceImpl.java

```diff
@@ -25,70 +37,146 @@
+		return studyInstance.isPresent() ? Optional.of(mapper.map(studyInstance.get(), StudyInstance.class)) : Optional.empty();
+	}
+
+	@Override
+	public InstanceData addInstanceData(final Integer studyId, final Integer instanceId, final InstanceData instanceData) {
+		this.studyValidator.validate(studyId, true);
+		this.instanceValidator.validateStudyInstance(studyId, Collections.singleton(instanceId));
+
+		final Integer datasetId = this.getEnvironmentDatasetId(studyId);
+		final Integer variableId = instanceData.getVariableId();
+		this.datasetValidator.validateExistingDatasetVariables(studyId, datasetId, Collections.singletonList(
+			variableId));
+		this.observationValidator.validateObservationValue(variableId, instanceData.getValue());
+
+		instanceData.setInstanceId(instanceId);
+		final boolean isEnvironmentCondition =
+			this.datasetService.getDatasetVariablesByType(studyId, datasetId, VariableType.STUDY_CONDITION).stream()
+				.anyMatch(v -> v.getId().equals(variableId));
+		this.middlewareStudyInstanceService.addInstanceData(instanceData, isEnvironmentCondition);
+		return instanceData;
+	}
+
+	@Override
+	public InstanceData updateInstanceData(final Integer studyId, final Integer instanceId, final Integer instanceDataId,
+		final InstanceData instanceData) {
+		this.studyValidator.validate(studyId, true);
+		this.instanceValidator.validateStudyInstance(studyId, Collections.singleton(instanceId));
+
+		final Integer datasetId = this.getEnvironmentDatasetId(studyId);
+		final Integer variableId = instanceData.getVariableId();
+		this.datasetValidator.validateExistingDatasetVariables(studyId, datasetId, Collections.singletonList(
+			variableId));
+		this.observationValidator.validateObservationValue(variableId, instanceData.getValue());
+
+		final boolean isEnvironmentCondition =
+			this.datasetService.getDatasetVariablesByType(studyId, datasetId, VariableType.STUDY_CONDITION).stream()
+				.anyMatch(v -> v.getId().equals(variableId));
+		final BindingResult errors = new MapBindingResult(new HashMap<String, String>(), Integer.class.getName());
+		final Optional<InstanceData> existingEnvironmentData =
+			this.middlewareStudyInstanceService.getInstanceData(instanceId, instanceDataId, variableId, isEnvironmentCondition);
+		if (!existingEnvironmentData.isPresent()) {
+			errors.reject(INVALID_ENVIRONMENT_DATA_ID);
+		} else if (!existingEnvironmentData.get().getVariableId().equals(variableId)) {
+			errors.reject(INVALID_VARIABLE_FOR_ENVIRONMENT_DATA);
+		}
+
+		if (!errors.getAllErrors().isEmpty()) {
+			throw new ApiRequestValidationException(errors.getAllErrors());
+		}
+
+		instanceData.setInstanceDataId(instanceDataId);
+		instanceData.setInstanceId(instanceId);
+		this.middlewareStudyInstanceService.updateInstanceData(instanceData, isEnvironmentCondition);
+		return instanceData;
+	}
+
+	private Integer getEnvironmentDatasetId(final Integer studyId) {
```

Comment:
@darla-leafnode This function should be partd of StudyService, it is the study responding by the datasets that it has.